### PR TITLE
Feature/refactor testrunner

### PR DESCRIPTION
**Pull request by dothebart**
### Scope & Purpose

`tu.performTests()` grew to become a massive beast.
 It is translated into the `testRunner` class now living in `js/client/modules/@arangodb/testutils/testrunner.js` which is the core of this PR.
 The class: 
 - replaces the `startStopHandlers` structure for plugins by abstract methods to be overwritten in deductions.
 - replaces the `runFn` method pointer argument by `runOneTest` abstract function
 - ready existing `runFn` functions were converted:
   - `tu.runThere` -> `tu.runOnArangodRunner` - spawn the test on the arangod / coordinator via .js
   - `tu.runInArangosh` -> `tu.runInArangoshRunner` - spawn an arangosh to launch the test in
   - `tu.runInLocalArangosh` -> `tu.runLocalInArangoshRunner` - eval the test into the current arangosh
 - driver tests even overwrite the `run` method, since they don't iterate over testfiles in first place.
 - The `restart` tests would contain an outdated copy of `tu.runInLocalArangosh` that would set the `instanceInfo` and the `options` to the `global` namespace. The PR fixes this WET programming by putting the new class on a global variable and changing the 3 test files to follow this additional change. 

All other files in `testsuites/` follow the API change of this kind and should not contain additional functional changes.

- [x] :hammer: Refactoring/simplification

### Checklist

- [x] Tests
  - [x] **integration tests**

### CODE REVIEW

File: js/client/modules/@arangodb/testutils/testrunner.js [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+    this.memProfCounter = 0;
+    this.env = {};
+    this.results = {};
+    this.continueTesting = true;
+    this.usersCount = 0;
+    this.cleanupChecks = [ ];
+    let usersTests = {
+      name: 'users',
+      setUp: function (te) {
+        if ((testname !== 'agency') && (testname !== 'authentication')){
```
Comment (jsteemann) [inline js/client/modules/@arangodb/testutils/testrunner.js:70]:
```suggestion
        if (testname !== 'agency' && testname !== 'authentication') {
```

File: js/client/modules/@arangodb/testutils/test-utils.js [hunk 2]
```diff
@@ -632,11 +200,11 @@ function splitBuckets (options, cases) {
     return cases;
   }
   if (cases.length === 0) {
-    didSplitBuckets = true;
+    didSplitBuckets(true);
     return cases;
   }
 
-  didSplitBuckets = true;
+  didSplitBuckets(didSplitBuckets);
```
Comment (jsteemann) [inline js/client/modules/@arangodb/testutils/test-utils.js:207]: This line of code puzzles me a bit. What kind of magic is going on here?

Comment (dothebart) [reply js/client/modules/@arangodb/testutils/test-utils.js:207]: its moving the location of this static variable.

File: js/client/modules/@arangodb/testutils/testrunner.js [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+          ' ] - Original test status: ' +
+          JSON.stringify(obj.results[te])
+      };
+      return false;
+    }
+  }
+};
+
+class testRunner {
+  constructor(options, testname, serverOptions, checkUsers=true, checkCollections=true) {
```
Comment (adityamukho) [inline js/client/modules/@arangodb/testutils/testrunner.js:251]:
```suggestion
  constructor(options, testname, serverOptions = {}, checkUsers=true, checkCollections=true) {
```

File: js/client/modules/@arangodb/testsuites/endpoints.js [hunk 4]
```diff
@@ -51,248 +51,259 @@ const testPaths = {
   'endpoints': [tu.pathForTesting('client/endpoint-spec.js')]
 };
 
-function endpoints (options) {
-  print(CYAN + 'Endpoints tests...' + RESET);
+class importRunner extends tu.runInArangoshRunner {
```
Comment (adityamukho) [inline js/client/modules/@arangodb/testsuites/endpoints.js:54]:
```suggestion
class endpointRunner extends tu.runInArangoshRunner {
```

File: js/client/modules/@arangodb/testsuites/endpoints.js [hunk 5]
```diff
@@ -51,248 +51,259 @@ const testPaths = {
...
-    return results;
-  }, { failed: 0, shutdown: true });
+      return results;
+    }, { failed: 0, shutdown: true });
+  }
 }
 
+function endpoints (options) {
+  print(CYAN + 'Endpoints tests...' + RESET);
+  return new importRunner(options, "endpoints").run();
```
Comment (adityamukho) [inline js/client/modules/@arangodb/testsuites/endpoints.js:305]:
```suggestion
  return new endpointRunner(options, "endpoints").run();
```

File: js/client/modules/@arangodb/testsuites/replication.js [hunk 6]
```diff
@@ -87,92 +87,100 @@ function shellClientReplicationApi (options) {
...
 
-// //////////////////////////////////////////////////////////////////////////////
-// / @brief TEST: replication_fuzz
-// //////////////////////////////////////////////////////////////////////////////
 
-function replicationFuzz (options) {
-  let testCases = tu.scanTestPaths(testPaths.replication_fuzz, options);
+class replicationRunner extends tu.runInArangoshRunner {
+  constructor(options, testname, serverOptions, startReplication=false) {
+    super(options, testname, serverOptions, true);
```
Comment (adityamukho) [inline js/client/modules/@arangodb/testsuites/replication.js:102]:
```suggestion
    super(options, testname, serverOptions);
```

File: js/client/modules/@arangodb/testsuites/rta_makedata.js [hunk 7]
```diff
@@ -66,71 +67,82 @@ const testPaths = {
...
+      super(options, testname, serverOptions, checkUsers, checkCollections);
+      this.info = "runRtaInArangosh";
+    }
+    runOneTest(file) {
+      let res = {'total':0, 'duration':0.0, 'status':true};
+      let tests = [
         fs.join(options.rtasource, 'test_data', 'makedata.js'),
         fs.join(options.rtasource, 'test_data', 'checkdata.js'),
         fs.join(options.rtasource, 'test_data', 'checkdata.js'),
         fs.join(options.rtasource, 'test_data', 'cleardata.js'),
```
Comment (adityamukho) [inline js/client/modules/@arangodb/testsuites/rta_makedata.js:81]:
```suggestion
        fs.join(this.options.rtasource, 'test_data', 'makedata.js'),
        fs.join(this.options.rtasource, 'test_data', 'checkdata.js'),
        fs.join(this.options.rtasource, 'test_data', 'checkdata.js'),
        fs.join(this.options.rtasource, 'test_data', 'cleardata.js'),
```

File: js/client/modules/@arangodb/testsuites/rta_makedata.js [hunk 8]
```diff
@@ -66,71 +67,82 @@ const testPaths = {
...
+        if (this.addArgs !== undefined) {
+          args = Object.assign(args, this.addArgs);
+        }
+        let argv = toArgv(args);
+        argv = argv.concat(['--',
+                            '--minReplicationFactor', '2',
+                            '--progress', 'true',
+                            '--oldVersion', db._version()
+                           ]);
+        if (options.hasOwnProperty('makedata_args')) {
```
Comment (adityamukho) [inline js/client/modules/@arangodb/testsuites/rta_makedata.js:105]:
```suggestion
        if (this.options.hasOwnProperty('makedata_args')) {
```

**Review by adityamukho — CHANGES_REQUESTED** [review]
Some changes requested.

File: js/client/modules/@arangodb/testsuites/dump.js [hunk 9]
```diff
@@ -374,21 +374,24 @@ class DumpRestoreHelper {
...
+    this.results.test = this.runOneTest(file);
+    this.addArgs = undefined;
     return this.validate(this.results.test);
   }
 
   runReTests(file, database) {
     this.print('revalidating modifications - ' + file);
-    db._useDatabase(database);
-    this.results.test = this.arangosh(file, {'server.database': database});
+    this.addArgs = {'server.database': database};
```
[the remark this answers is not in the document]
Comment (dothebart) [reply js/client/modules/@arangodb/testsuites/dump.js:385]: We derive here:
```
class DumpRestoreHelper extends tu.runInArangoshRunner {
```
so `this.runOneTest` spawns a subshell; `db._useDatabase(database);` to change the database of the local arangosh doesn't have an effect on these.

File: js/client/modules/@arangodb/testsuites/java.js [hunk 10]
```diff
@@ -65,179 +66,183 @@ const testPaths = {
...
+      if (this.options.testCase) {
+        args.push('-Dtest=' + this.options.testCase);
+        args.push('-DfailIfNoTests=false'); // if we don't specify this, errors will occur.
+      }
+      if (this.options.hasOwnProperty('javaOptions')) {
+        for (var key in this.options.javaOptions) {
+          args.push('-D' + key + '=' + this.options.javaOptions[key]);
+        }
+      }
+      if (this.options.extremeVerbosity || true) {
```
[the remark this answers is not in the document]
Comment (dothebart) [reply js/client/modules/@arangodb/testsuites/java.js]: removed, tnx.

File: js/client/modules/@arangodb/testsuites/backup.js [hunk 11]
```diff
@@ -76,137 +76,150 @@ const failPreStartMessage = (msg) => {
...
+  // //////////////////////////////////////////////////////////////////////////////
+  generateDumpData() {
+    let options = _.clone(this.options);
+    if (this.dumpPath !== undefined) {
+      return dumpPath;
+    }
+    
+    const conf = {
+      'server.authentication': 'true',
+      'server.jwt-secret': 'haxxmann'
```
[the remark this answers is not in the document]
Comment (dothebart) [reply js/client/modules/@arangodb/testsuites/backup.js]: https://github.com/arangodb/arangodb/pull/15998/commits/41062aac84cff06024c76558a5e6e910ac57f8b3 does this.

**Review by adityamukho — APPROVED** [review]
LGTM.

**Review by KVS85 — APPROVED** [review]
LGTM.

**Pull request merged**